src/main/java/org/wildfly/security/password/spec/BasicPasswordSpecEncoding.java

```diff
@@ -0,0 +1,190 @@
+import org.wildfly.security.password.Password;
+import org.wildfly.security.password.PasswordFactory;
+import org.wildfly.security.password.interfaces.ClearPassword;
+import org.wildfly.security.password.interfaces.DigestPassword;
+import org.wildfly.security.util.ByteIterator;
+import org.wildfly.security.util.ByteStringBuilder;
+import org.wildfly.security.util.CodePointIterator;
+
+import java.nio.charset.StandardCharsets;
+import java.security.NoSuchAlgorithmException;
+import java.security.spec.InvalidKeySpecException;
+
+/**
+ * Provide methods for encoding and decoding of {@link PasswordSpec}.
+ *
+ * @author <a href="mailto:david.lloyd@redhat.com">David M. Lloyd</a>
+ * @author <a href="mailto:psilva@redhat.com">Pedro Igor</a>
+ */
+public final class BasicPasswordSpecEncoding {
+
+    // the value for each identifier must be kept and can not change
+    private static final byte CLEAR_PASSWORD_SPEC_ID = 1;
+    private static final byte DIGEST_PASSWORD_SPEC_ID = 2;
+    private static final byte HASH_PASSWORD_SPEC_ID = 3;
+    private static final byte SALTED_HASH_PASSWORD_SPEC_ID = 4;
+    private static final byte ITERATED_SALTED_HASH_SPEC_ID = 5;
+
+    private BasicPasswordSpecEncoding() {}
+
+    /**
+     * Encode the given {@link PasswordSpec} to a byte array.
+     *
+     * @param password the password to encode
+     * @return a byte array representing the encoded password or null if no encoder was capable to encode the given password
+     */
+    public static byte[] encode(PasswordSpec passwordSpec) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        if (ClearPasswordSpec.class.isInstance(passwordSpec)) {
+            return encodeClearPasswordSpec((ClearPasswordSpec) passwordSpec);
+        } else if (DigestPasswordSpec.class.isInstance(passwordSpec)) {
+            return encodeDigestPasswordSpec((DigestPasswordSpec) passwordSpec);
+        } else if (SaltedHashPasswordSpec.class.isInstance(passwordSpec)) {
+            return encodeSaltedHashPasswordSpec((SaltedHashPasswordSpec) passwordSpec);
+        } else if (IteratedSaltedHashPasswordSpec.class.isInstance(passwordSpec)) {
+            return encodeIteratedSaltedHashSpec((IteratedSaltedHashPasswordSpec) passwordSpec);
+        } else if (HashPasswordSpec.class.isInstance(passwordSpec)) {
+            return encodeHashPasswordSpec((HashPasswordSpec) passwordSpec);
+        }
+
+        return null;
+    }
+
+    /**
+     * Encode the given {@link Password} to a byte array.
+     *
+     * @param password the password to encode
+     * @return a byte array representing the encoded password or null if no encoder was capable to encode the given password
+     */
+    public static byte[] encode(Password password) throws NoSuchAlgorithmException, InvalidKeySpecException {
+        PasswordFactory passwordFactory = PasswordFactory.getInstance(password.getAlgorithm());
+
+        if (ClearPassword.class.isInstance(password)) {
```

Comment:
See my note on the previous diff here: the passwordFactory should be queried using the ``convertibleToKeySpec`` method instead of ``instanceof`` checks.